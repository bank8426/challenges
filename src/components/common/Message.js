import React from 'react'
import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types'

/** make StyledMessageList alway display at top right of the screen.*/
const StyledMessageList = styled.div`
  position: fixed;
  right: 0;
  top: 0;
`;
/** 
 * keyframes to make message move into the screen from the right 
 * and then display for a while before disappear
 * */
const messageInKeyframe = keyframes`
  0% {
    -webkit-transform: translateX(100%);
    transform: translateX(100%);
  }
  20% {
    -webkit-transform: translateX(0);
    transform: translateX(0);
  }
  80% {
    opacity:1;
  }
  100% {
    opacity:0;
  }
`

const StyledMessage = styled.p`
  /** change message background color follow isErrorMessage*/
  background: ${props => props.isErrorMessage ? '#EB5757' : '#2F80ED' };
  color: white;
  margin: 0 1em 0;
  padding:10px;
  font-weight: bold;
  text-align: center;
  box-shadow: 2px 2px 10px rgba(0, 0, 0, 0.25);
  visibility: visible;
  overflow: hidden;
  -webkit-animation: ${messageInKeyframe} 2s forwards cubic-bezier(0.2, 0.8, 0.2, 1);
  animation: ${messageInKeyframe} 2s forwards cubic-bezier(0.2, 0.8, 0.2, 1);

  /** add margin top for first child */
  &:nth-child(1) {
    margin-top: 1em;
  }
`;

/** div to display multiple messages */
const Message = ({messages}) => {
  return (
    <StyledMessageList>
      { 
        messages.map(({id,message,isErrorMessage}) => (
          <StyledMessage key={id} isErrorMessage={isErrorMessage}>
            {message}
          </StyledMessage>
        ))
      }
    </StyledMessageList>
  )
}

Message.defaultProps = {
  messages: [],
}

Message.propTypes={
  /** array of messages to be display.*/
  messages: PropTypes.arrayOf(PropTypes.shape({
    /** id of message*/
    id :PropTypes.number.isRequired,
    /** message string to display*/
    message:PropTypes.string.isRequired,
    /** boolean indicate message type*/
    isErrorMessage:PropTypes.bool.isRequired,
  })),
}

export default Message
