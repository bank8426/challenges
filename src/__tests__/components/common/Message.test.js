import React from 'react'
import Message from '../../../components/common/Message'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

it('should display message', () => {
  const tree = renderer.create(<Message />).toJSON()
  expect(tree).toMatchSnapshot()
})


it('should display red message for error message', () => {
  const initMessages =  [{id : 1,message: 'Error',isErrorMessage :true}]
  const tree = renderer.create(<Message messages={initMessages} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should display blue message for normal message', () => {
  const initMessages =  [{id : 1,message: 'Thank you',isErrorMessage :false}]
  const tree = renderer.create(<Message messages={initMessages} />).toJSON()
  expect(tree).toMatchSnapshot()
})