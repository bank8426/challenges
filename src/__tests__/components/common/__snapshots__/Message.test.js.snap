// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display blue message for normal message 1`] = `
.c0 {
  position: fixed;
  right: 0;
  top: 0;
}

.c1 {
  background: #2F80ED;
  color: white;
  margin: 0 1em 0;
  padding: 10px;
  font-weight: bold;
  text-align: center;
  box-shadow: 2px 2px 10px rgba(0,0,0,0.25);
  visibility: visible;
  overflow: hidden;
  -webkit-animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
  -webkit-animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
  animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
}

.c1:nth-child(1) {
  margin-top: 1em;
}

<div
  className="c0"
>
  <p
    className="c1"
  >
    Thank you
  </p>
</div>
`;

exports[`should display message 1`] = `
.c0 {
  position: fixed;
  right: 0;
  top: 0;
}

<div
  className="c0"
/>
`;

exports[`should display red message for error message 1`] = `
.c0 {
  position: fixed;
  right: 0;
  top: 0;
}

.c1 {
  background: #EB5757;
  color: white;
  margin: 0 1em 0;
  padding: 10px;
  font-weight: bold;
  text-align: center;
  box-shadow: 2px 2px 10px rgba(0,0,0,0.25);
  visibility: visible;
  overflow: hidden;
  -webkit-animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
  -webkit-animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
  animation: bfJSmJ 2s forwards cubic-bezier(0.2,0.8,0.2,1);
}

.c1:nth-child(1) {
  margin-top: 1em;
}

<div
  className="c0"
>
  <p
    className="c1"
  >
    Error
  </p>
</div>
`;
